server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -328,6 +345,100 @@ void addNodes(int newNodesCount) {
+                    deterministicTaskQueue.setExecutionDelayVariabilityMillis(DEFAULT_DELAY_VARIABILITY);
+                    logger.debug("----> [runRandomly {}] reducing delay variability and running until [{}]", step, finishTime);
+                }
+
+                try {
+                    if (rarely()) {
+                        final ClusterNode clusterNode = getAnyNodePreferringLeaders();
+                        final int newValue = randomInt();
+                        logger.debug("----> [runRandomly {}] proposing new value [{}] to [{}]", step, newValue, clusterNode.getId());
+                        clusterNode.submitValue(newValue);
+                    } else if (rarely()) {
+                        final ClusterNode clusterNode = getAnyNode();
+                        logger.debug("----> [runRandomly {}] forcing {} to become candidate", step, clusterNode.getId());
+                        synchronized (clusterNode.coordinator.mutex) {
+                            clusterNode.coordinator.becomeCandidate("runRandomly");
+                        }
+                    } else if (rarely()) {
+                        final ClusterNode clusterNode = getAnyNode();
+                        final String id = clusterNode.getId();
+                        disconnectedNodes.remove(id);
+                        blackholedNodes.remove(id);
+
+                        switch (randomInt(2)) {
+                            case 0:
+                                logger.debug("----> [runRandomly {}] connecting {}", step, id);
+                                break;
+                            case 1:
+                                logger.debug("----> [runRandomly {}] disconnecting {}", step, id);
+                                disconnectedNodes.add(id);
+                                break;
+                            case 2:
+                                logger.debug("----> [runRandomly {}] blackholing {}", step, id);
+                                blackholedNodes.add(id);
+                                break;
+                        }
+                    } else {
+                        if (deterministicTaskQueue.hasDeferredTasks() && randomBoolean()) {
+                            deterministicTaskQueue.advanceTime();
+                        } else if (deterministicTaskQueue.hasRunnableTasks()) {
+                            deterministicTaskQueue.runRandomTask();
+                        }
+                    }
+
+                    // TODO other random steps:
+                    // - reboot a node
+                    // - abdicate leadership
+                    // - bootstrap
+
+                } catch (CoordinationStateRejectedException ignored) {
+                    // This is ok: it just means a message couldn't currently be handled.
+                }
+
+                assertConsistentStates();
+            }
+
+            disconnectedNodes.clear();
```

Comment:
I think running randomly with disruptions already in place needs more thought since it might be unexpected to connect a node that the caller explicitly disconnected. I think the best thing to do here is to refuse to run in such circumstances, and then if we really need it we can revisit this later. I pushed af227e723288e9c0443ca0a756fff530d9c6a24b.